### PR TITLE
feat(textfield): update component to meet new specifications

### PR DESCRIPTION
**Pull request by lauralucca**
# Description

This updates the **TextField** component, including new features and visual updates, and deals with technical and functional debts, as described in [DSY-2081 Update | TextField [RN]](https://natura.atlassian.net/browse/DSY-2081), and listed below:

- [x] Adding surface color background to the field that was previously missing;
- [x] Updating `Read Only` styles and behavior to better fit design definitions;
- [x] Adding new icons to helper texts when the component show visual feedback for validation **error** or **success** for accessibility purposes;
- [x] Updating `size` names to fit new design definitions (previous used sizes are now `DEPRECATED`);
- [x] Adding `action` attribute that was not developed before;
- [x] Adding `number` option in attribute `type` that was not developed before;
- [x] Fixing placeholder not showing properly for size `medium` bug on Android;
- [x] Refactoring code for good practices and reusability extracting components `InputBox`, `InputFeedbackContainer`, `InputHelperText`, `InputLabel`;
- [x] Updating documentation with new features.

No dependencies were added or removed for this change.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

To see the visual changes included in this, you may want to run the sample phone app or the storybook web app and check under the `TextField` component section.

**There should an example for every utilization, attribute and state presented in the [UI documentation](https://xd.adobe.com/view/a33f9f09-27d3-4997-907d-dedf22c647dc-f66c/):**
- [x] A section called **"Default"** showing the component in its default state, without any customization;
- [x] A section called **"Sizes"** showing examples of the component in all available sizes: `medium` and `mediumX`;
- [x] A section called **"States"** showing examples of the component in its states `enabled` and `filled`;
- [x] A section called **"Feedback"** showing examples of the component showing visual feedback for valitation status `error` and `success`;
- [x] A section called **"Required"** showing the component with the `required` attribute applied;
- [x] A section called **"Disabled"** showing the component with the `disabled` attribute applied;
- [x] A section called **"Readonly"** showing the component with the `readonly` attribute applied;
- [x] A section called **"Action"** showing examples of the component with an `IconButton` and an `Image` component inside;
- [x] A section called **"Type"** showing examples of the component in all available types: `text`, `password` and `number`;
- [x] A section called **"All"** showing all the examples above.

**To ensure the correct the behaviour and provide reliability the following automated tests were included and/or updated:**

- [src/common/themeSelectors/opacity/opacity.test.ts](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-898023b71982f630e88158a2933b2baa8d33077bdbcb7b2e21d77474fe3a9831)
- [src/common/themeSelectors/sizes/sizes.test.ts](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-ad20d987999f334535f475d87a341b4ea89afa3ae8f91ed3ff541a77618d9b31)
- [src/common/themeSelectors/typography/typography.test.ts](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-e5492e5a7052c029f89e0031b0e53e0b46820f7702829bde6473d324b5ed6645)
- [src/components/InputFeedbackContainer/InputFeedbackHelper.test.ts](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-f09b3f8a783b92905653e04f42eefadd378bcec263d2636313db7adbd1b97a27)
- [src/components/InputBox/InputBox.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-0378d81f7c35f5cc0775c1520dc9c462890cc72fed2deaf50e4d8c8f156c3673)
- [src/components/InputFeedbackContainer/InputFeedbackContainer.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-9bc4d198fd2535cb1156b25eb2141168569dfb0f66ca0ddf01c5c8414fada9e9)
- [src/components/InputHelperText/InputHelperText.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-377cd4b523641aa5bd62edebd56d58412a97da8182b88a5adcf2e42ccf781f8d)
- [src/components/InputLabel/InputLabel.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-ce2667a5bc2632235ee4d3668ad3dc71c9d45f54870472d20ee99e36592cd628)
- [src/components/TextField/TextField.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/177/files#diff-e972727840e92994e251e65611d1b647350c2fb7205da936ad192c7a7887c5ad)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/common/themeSelectors/sizes/sizes.ts [hunk 1]
```diff
@@ -5,8 +5,12 @@ const getRadiusSizes = (theme: Theme) => checkTheme(theme).borderRadius;
 
 export const getRadiusBySize = (theme: Theme, size: keyof Size) => getRadiusSizes(theme)[size];
 
+export const getBorderRadiusMedium = (theme: Theme) => getRadiusBySize(theme, 'medium');
```
Comment (arielwb) [inline src/common/themeSelectors/sizes/sizes.ts:8]: faltam os testes das novas funções desse arquivo

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+}: InputFeedbackContainerProps, theme: Theme) => {
+  if (disabled) return getColorLowEmphasis(theme);
+  if (feedback === 'success') return getColorSuccess(theme);
+  if (feedback === 'error') return getColorAlert(theme);
+
+  return getColorMediumEmphasis(theme);
+};
+
+/* eslint-disable complexity */
+const getBoxColor = ({
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:53]: refatorar para diminuir a complexidade

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  }
+  | {
+    active?: never;
+    disabled?: boolean;
+    feedback?: never;
+  }
+
+export type InputFeedbackContainerProps = ContentProps & FeedbackProps
+
+const getTextElementsColor = ({
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:42]: refatorar pra diminuir a complexidade

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  if (active) return getColorPrimary(theme);
+  if (feedback === 'success') return getColorSuccess(theme);
+  if (feedback === 'error') return getColorAlert(theme);
+  if (filled) return getColorHighEmphasis(theme);
+
+  return getColorLowEmphasis(theme);
+};
+/* eslint-enable complexity */
+
+const HelperText = ({ children, color, feedback }) => {
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:66]: separa os sub componentes e suas funções em arquivos pra facilitar a leitura.
aqui temos logica e tipos da label, helper text e input todos juntos

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  const theme = useTheme();
+  const iconName = (feedback === 'success' ? 'outlined-action-check' : 'outlined-action-cancel');
+  const iconColor = (feedback === 'success' ? 'success' : 'alert');
+  const { caption } = getTypographyStyles(theme);
+
+  return (
+    <View style={{ width: '100%' }}>
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:73]: eu entendo que é chato fazer um styled componente pra cada VIew, mas o resultado final desse componente me pareceu mais dificil de ler do que necessario. Talvez so extraindo o componente do icone já ajude, mas acho que da pra melhorar

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  } = props;
+  const textElementsColor = getTextElementsColor(props, theme);
+  const boxColor = getBoxColor(props, theme);
+
+  return (
+    <View style={{ width: '100%' }}>
+      { label
+        && <Label color={textElementsColor} required={required}>{label}</Label>
+      }
+      <View style={{
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:128]: criar styled components pra essas views

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        style={{
+          flexDirection: 'row',
+          marginTop: getSpacingMicro(theme),
+        }}
+      >
+        { feedback
+          && <View style={{
+            marginRight: getSpacingMicro(theme),
+          }}>
+            <Icon testID="icon" name={iconName} color={iconColor} size="small" />
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:84]: o icone do feedback tem ação no click? por ex, a usuaria poderia inserir um icone de olho e querer mostrar o password no click

File: src/components/TextField/TextField.tsx [hunk 8]
```diff
@@ -1,243 +1,217 @@
...
+    selectTextOnFocus,
+    selection,
+    selectionState,
+    showSoftInputOnFocus,
+    size = 'mediumX',
+    spellCheck,
+    textBreakStrategy,
+    textContentType,
+    value = '',
+  }: TextFieldProps = props;
```
Comment (arielwb) [inline src/components/TextField/TextField.tsx:91]: eu acho mto ruim dar o spread de todas as props assim e ter que repassar todas pro componente de TextInput.
sera que nao enocntramos uma forma melhor de resolver isso?

File: src/components/TextField/TextField.tsx [hunk 9]
```diff
@@ -1,243 +1,217 @@
...
-        size={size}
+    <InputFeedbackContainer
+      filled={filled}
+      helperText={helperText}
+      label={label}
+      required={required}
+      {...getContainerProps()}
+    >
+      <TextInput testID="input"
+        style={{
```
Comment (arielwb) [inline src/components/TextField/TextField.tsx:130]: styled component

File: src/components/TextField/TextField.tsx [hunk 10]
```diff
@@ -1,243 +1,217 @@
...
+        selection={selection}
+        selectionState={selectionState}
+        showSoftInputOnFocus={showSoftInputOnFocus}
+        spellCheck={spellCheck}
+        textBreakStrategy={textBreakStrategy}
+        textContentType={textContentType}
+        value={value}
+      />
+      { action === 'icon' && actionOnPress
+        && <View style={{ paddingRight: getSpacingTiny(theme) }}>
```
Comment (arielwb) [inline src/components/TextField/TextField.tsx:192]: extrair isso em um componente

File: src/components/TextField/TextField.tsx [hunk 11]
```diff
@@ -1,243 +1,217 @@
...
+            <IconButton testID='action-icon' iconColor={iconColor} icon={iconName} onPress={actionOnPress} />
+          </View>
+      }
+      { action === 'image' && imageSource
+        && <TouchableRipple
```
Comment (arielwb) [inline src/components/TextField/TextField.tsx:197]: extrair isso em um componente

**Review by arielwb — CHANGES_REQUESTED** [review]

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx (hunk 7 above)
Comment (lauralucca) [reply src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:84]: Nesse caso acho que você tá pensando no ícone de ação do TextField. Esse aqui é um íconezinho de indicação que aparece embaixo do campo quando muda a cor (para quem não diferencia vermelho ou verde). É o (𐄂) e o (✓) na imagem abaixo:

image

Com a refatoração desse componente e separação de tudo vou deixar mais nítido o que representa o quê!

File: src/components/TextField/TextField.types.ts [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+   */
+  helperText: string;
+}
+
+interface WithoutFeedbackProps {
+  feedback?: never;
+  helperText?: string;
+}
+
+interface ActionImageProps {
```
[the remark this answers is not in the document]
Comment (lauralucca) [reply src/components/TextField/TextField.types.ts:145]: Conversei com o Rafa sobre isso, ele respondeu que imaginou a imagem com ação sim, opcional. Eu tinha deduzido isso de ele ter chamado o atributo "Action", mas eu acho que é questionável, se a gente entender como desnecessário. Parece que rola um YAGNI aqui.

[the remark this answers is not in the document]
Comment (lauralucca) [reply src/components/TextField/TextField.types.ts:145]: Removi o ripple e o onPress, para ficar igual as outras stacks. Vamos conversar com Rafa e Lucas sobre a necessidade de incluir isso.

File: src/components/TextField/TextField.tsx (hunk 8 above)
Comment (lauralucca) [reply src/components/TextField/TextField.tsx:48]: Testei de várias formas e a maioria deixa possível alterar as props de estilos que queremos evitar. Tenho essas duas sugestões que funcionam: https://github.com/natura-cosmeticos/natds-rn/pull/177/commits/c7801de3cfa89caea8f1d05567a5357a6c5f13ca (nesse commit eu substituí uma pela outra, então você pode ver uma em vermelho a outra em verde. Caso você veja algumas das duas como mais legível ou desejável, é só dar um toque)

Enquanto mexia nisso, fiz rapidinho um teste automatizado para garantir (embora não commitei ele, porque não achei que fazia sentido ter ele na lib) e confirmei que tudo funciona como esperado.

image

Percebi que outra opção seria passar as props que eu não quero permitir, que são menos, porque as que eu defini no componente a usuária não consegue sobrescrever. Mas essa última opção eu achei muito provável de criar confusão e de piorar o autocomplete do ts.

image

File: package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@naturacosmeticos/natds-rn",
-  "version": "7.0.0",
+  "version": "7.1.0-DSY-2081.4",
```
[the remark this answers is not in the document]
Comment (lauralucca) [reply package.json]: Realmente! A lógica anterior era muito complexa e acoplada, então eu precisei remover para desenvolver o atributo Action. Quando a gente conversou sobre os tipos Number e Multiline serem props do React Native, acabei deixando sem o Password também.

Eu prefiro evitar a breaking change, mas também acho que manter isso no componente deixa acoplado o `IconButton` com os ícones de 👁 e a prop de `'secureText'` e isso pode impedir a usuária de usá-las separadas.

@yasminnvaz @arielwb 
Se a gente estiver OK com isso, adiciono agora no componente!

Só para mostrar, também é bem simples de fazer pelo lado da usuária, dá uma olhada:

image

[the remark this answers is not in the document]
Comment (lauralucca) [reply package.json]: https://natds-rn.natura.design/DSY-2081-textfield/index.html?path=/story/components-textfield--password

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+  const colors = {
+    active: { box: getColorPrimary(theme), text: getColorMediumEmphasis(theme) },
+    default: { box: getColorLowEmphasis(theme), text: getColorMediumEmphasis(theme) },
+    disabled: { box: getColorLowEmphasis(theme), text: getColorLowEmphasis(theme) },
+    error: { box: getColorAlert(theme), text: getColorAlert(theme) },
+    filled: { box: getColorHighEmphasis(theme), text: getColorMediumEmphasis(theme) },
+    success: { box: getColorSuccess(theme), text: getColorSuccess(theme) },
+  };
+
+  return colors[state || 'default'][element];
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:42]: LINDOOOOOOOOO! <3

File: src/components/InputFeedbackContainer/InputFeedbackHelper.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+export interface InputFeedbackProps {
+  active?: boolean,
+  disabled?: boolean,
+  filled?: boolean,
+  feedback?: 'success' | 'error',
+}
+
+export const buildInputState = ({
+  active, disabled, feedback, filled,
+}: InputFeedbackProps) => {
```
Comment (arielwb) [inline src/components/InputFeedbackContainer/InputFeedbackHelper.ts:12]: aqui eu fiquei com duvidas:
essa função é usada somente pra pegar o estado correto pra usar na função `getElementsColorsByState`, certo?
Se sim, tenho alguns pontos:
- Acho que o retorno no if feedback deveria ser uma string, e nao um boolean
- feedback não existe em `getElementsColorsByState`
- error não existe em `buildInputState`
Acho que tu pode usar o typescript pra garantir que essas duas funções vao seguir alinhadas, dizendo que o tipo do retorno aqui é um `InputStates`. Hoje o retorno dessa função é string | boolean

File: src/components/InputFeedbackContainer/InputFeedbackContainer.tsx (hunk 14 above)
Comment (arielwb) [reply src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:42]: porem aqui no retorno acho que a gente pode remover o default, pq tu já retorna ele na `buildInputState`.
outra dica pra deixar tudo ainda mais amarradinho: acho que a const colors merece um tipo que garanta que as chaves desse objeto sejam somente do tipo `InputStates`.
Ai se um dia default virar banana, o TS vai te ajudar a saber onde tem que trocar

Comment (lauralucca) [reply src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:42]: Mas o `InputStates` já é um tipo, eu declarei na primeira linha do `./InputFeedbackHelper.ts`
Talvez faz mais sentido declarar aqui?

Comment (lauralucca) [reply src/components/InputFeedbackContainer/InputFeedbackContainer.tsx:42]: AH eu tinha entendido errado. Fiz o tipo e removi o "default", tudo certo!

File: src/components/InputFeedbackContainer/InputFeedbackHelper.ts (hunk 15 above)
Comment (lauralucca) [reply src/components/InputFeedbackContainer/InputFeedbackHelper.ts:12]: Coloquei o tipo InputStates no retorno!

Sobre o resto das alterações, já funciona assim! Na verdade o tipo de Feedback não é um boolean, é o union das strings "error" e "success". Por isso que essa função manda "error" e "success" pra `'getElementsColorsByState'`

Eu pensei em colocar mais explicito, tipo:
```
if (feedback === "error") return "error";
if (feedback === "success") return "success";
```
maaaaaas aí é mais um if né. Cansada de ifs. hahaha

**Review by yasminnvaz — DISMISSED** [review]

**Pull request merged**